### DISCUSSION TITLE
RFC: Violentmonkey Sync Revamp

### DISCUSSION THREAD

**Discussion by gera2ld** [violentmonkey/violentmonkey#1529] [Ideas]
## Goals

- Improve conflicts handling
- Support multiple profiles
- Sync global settings
- Sync script settings and data

Note: This RFC is only for sync with third-party cloud storage services. We may enable sync with `browser.storage` API in the future but that will be a totally different approach.

### Lock File

- Check lock
- Put a lock file (ttl: 5min)
- Wait 5s and check lock again, halt in case of race condition
- Halt if client time is significantly different (more than 1min) from server time because it may affect our lock flow
- Renew the lock every minute
- Halt if a valid lock file with a different uuid exists

### Conflict Handling

- Fetch manifest and compare
  - Handle conflicts
    - Conflicts happen when changes are made to a device whose data is not up to date

### File Synchronization

- Upload new files (overwrite existing/obsolete ones)
- Update manifest
- Remove obsolete files
- Remove lock file

### Prerequisites

- A UUID for this sync transaction
- Sync service configuration

### Multiple Profiles

All files will be put under a directory with the name of the profile. So we can have multiple profiles under different directories.

Since everything is limited in the profile directory, we can sync different profiles at the same time.

### Lock File

The lock file is used to avoid multiple sync processes at one time. The file name is `violentmonkey.lock.json`.

```typescript
{
  uuid: UUID;
  /** identity of the instance used for debugging */
  label: string;
  /** ISO timestamp */
  expireAt: string;
}
```

### Manifest

The manifest contains mappings from a script to its blobs. The file name is `violentmonkey.manifest.json`.

```typescript
{
  /** ISO timestamp, last sync time */
  updatedAt: string;
  settings: Record<string, unknown>;
  scripts: Array<{
    namespace: string;
    name: string;
    /** ISO timestamp */
    updatedAt: string;
    settingsUpdatedAt: string;
    dataUpdatedAt: string;
    /**
     * We will generate a human-readable file name for better maintenance,
     * excluding the extension name.
     */
    fileName?: string;
    /**
     * If allowUpdate is true, ignore conflicts and
     * just keep the latest one.
     */
    allowUpdate: boolean;
  }>;
}
```

### Local Scripts

Additional fields are required for local scripts:

```typescript
{
  /** ISO timestamp */
  syncedAt?: string;
  updatedAt: string;
  settingsUpdatedAt: string;
  dataUpdatedAt: string;
}
```

### Files

We generate a human-readable and unique file name for each script and store it in the manifest.

Each script will have three files in storage at most:

- script content is stored in `${fileName}.user.js`
- script settings are stored in `${fileName}.settings.json`
- script data is stored in `${fileName}.data.json`

## Compatibility

We will keep the old version of sync service for a while.

It shouldn't be a problem since the old version uses the root directory while the new version always use a subdirectory.

The only concern is that it's not possible to sync between old version and new version.

Anyway it's better to back up the files before switching to the new sync service.

## Cloud Services

We will continue to support Dropbox, OneDrive, Google Drive and WebDAV.

### Google Drive

Google Drive has two different modes.

The first mode is for web applications specifically, we store data in a separate space where users cannot access. The other mode is to get access to the whole drive and store files in a normal directory visible to its user.

Violentmonkey is using the first mode which confused some users because they cannot find their files.

However the second mode requires quite a large scope that it can literally access all files in a user's Google Drive. It may scare away users who are worried about privacy.

So **we will continue to use the application specific space** but we may provide a simple web-based file manager to help users manage their files under Violentmonkey. This file manager should be a separate web app and may also be able to manage files for other cloud services too.

### Lack of Version Control

- We don't store old versions of data, which means the sync process is irreversible. Even if a transaction is halted for whatever reason, **the processed files will be new and the unprocessed files will stay old**.

### Script Data

- All devices that have script data sync enabled will **share the same data for each script**.

**Comment by tophf** [comment]
1. Use a separate files for data and settings so we have three files per script. This would allow a simpler selective sync where the user on this particular device wants to limit the operation for whatever reason e.g. `[x] Script data` to allow the user to exclude it from sync on this particular device.
2. Must support the old data format at least to import the data and probably display a confirmation before upgrading it
3. Regarding "Lack of Version Control", maybe upload as `tmp` file, then rename the old files to `${oldName}.tmp`, tmp to oldName, delete `${oldName}.tmp`.
4. Maybe zip each file except settings

**Reply by gera2ld** [reply]
> 1. Use a separate files for data and settings so we have three files per script. This would allow a simpler selective sync where the user on this particular device wants to limit the operation for whatever reason e.g. `[x] Script data` to allow the user to exclude it from sync on this particular device.

Reasonable, let me update.

> 2. Must support the old data format at least to import the data and probably display a confirmation before upgrading it

We can keep old service for a while, allowing users to sync from old data format. Once it's synced to local, they can switch to the new version and sync local data to the new service. That's why I said "it's not possible to sync between old version and new version" but should not be a problem.

The confirmation is required because the configuration might be different and need users to confirm.

> 3. Regarding "Lack of Version Control", maybe upload as `tmp` file, then rename the old files to `${oldName}.tmp`, tmp to oldName, delete `${oldName}.tmp`.

I think this is something different from what I meant. Even in this way, we may have some files renamed while some not, as a result, some files are new and some are still old.

If I understand correctly, this way can only prevent a file from being updated partially. I think a cloud storage service will do this when we send a *putFile* request, so maybe this is not necessary.

A possible approach is to allow multiple versions of files stored with different names like `fileName-1`, `fileName-2`, then after all files uploaded, we store the fileName with the correct revision into manifest and deprecate all those not referenced, so all versions will be updated at once. But in that way we waste a lot more storage and request counts which I think is not worth it.

> 4. Maybe zip each file except settings

I agree to zip files optionally, especially helpful for limited storage and poor network.

**Comment by da2x** [comment]
As mentioned in #1260, on marking a script as delete, you’d need to rename it to remove the user data from the file name. Choosing random file names (e.g, a UUID) avoids this complication.

**Reply by gera2ld** [reply]
The file name is just for maintenance because some people just want to see their scripts.
`namespace` + `name` is used as the unique identity of a script to do the comparing. There is a mapping from this unique identity to file name stored in manifest.

**Comment by da2x** [comment]
>> Maybe zip each file except settings
>
> I agree to zip files optionally, especially helpful for limited storage and poor network.

Individual compression is unlikely to yield much benefit; user scripts are too short to deflate well. During downloads, the server will most likely offer them with gzip compression (HTTP Content Negotiation). This handles the data compression needs during initial sync, which is the single biggest data transfer in the app’s lifetime. Uploads won’t be compressed, but how often does anyone upload new scripts?

To save space on the server, you could enable file system compression on the server. Again, the scripts are too small to benefit. They’re smaller than one extent/inode, so you don’t gain anything from compression.

The only way to benefit in on-disk storage is to compress multiple files together. However, this negates the benefits of reducing upload/download since you’d need to retransfer multiple files to reapply the compression on them. So, you‘d sacrifice bandwidth efficiency for space savings.

50 uncompressed scripts would use, what, like 100 kB? Compression could potentially cut 20–60% off that figure assuming they were all bundled together in one stream/file.

In conclusion, there’s very little benefit to zipping or compressing scripts in the client.

**Reply by gera2ld** [reply]
This is arguable. I know many scripts more than 1MB, and there are scripts with very large data. In either case compression may be helpful. I believe you don't use powerful usescripts much if 50 of them only sum up to 100kB.

Besides, under poor network file size matters a lot. For example in China using a cloud service from another country (e.g. OneDrive), it's very likely that transferring a file more than 1MB fails again and again but a 300KB file will succeed.

Anyway the benefit is limited, this can be deprioritized.
